<!DOCTYPE html>

<meta charset="utf-8">
<title>e!Peek API</title>

<style>
  @import url(../../../index.css);
  <!-- insert in-place styles here -->
</style>

<div id="body">
  <header>
    <script src="../../../lib/d3.min.js"></script>
    <a href="../../../index.html">Overview</a>
    <a style="margin-left:1em;" href="../../../themes/index.html">Themes</a>
    <a style="margin-left:1em;" href="./">Examples</a>
    <b><a style="margin-left:1em;" href="../index.html">Documentation</a></b>
    <a style="margin-left:1em;" href="https://github.com/emepyc">Source</a>
  </header>

  <h1>API Reference</h1>
  <a href="../index.html"><h2>Genome Browser</h2></a>

<!-- Starting -->
  <h2><a name="starting" href="#starting">#</a>Starting the plug-in</h2>
  <p>Starting the plug-in means starting the genome browser in the specified <i>DOM element</i>. This is normally done inside of the <a href="">theme</a> after everything has been configured. No <i>REST</i> query is done until this method is called. The plug-in can be safely re-started as many times as needed (for example as a result of a search functionality in the theme as in the <a href="../../../themes/compact/compact.html">compact theme</a>).</p>

  <p><a name="start" href="#start">#</a> genome_browser.<b>start</b> ([where])</p>
  <p>The start method starts the genome browser on the specified location. The location is represented as an object having one of these combinations of fields:</p>
  <ul>
    <li><i>species</i>, <i>chr</i>, <i>from</i>, <i>to</i></li>
    <li><i>species</i>, <i>gene name</i></li>
  </ul>
  <p>For example:</p>
  <pre><code class="javascript">
var genome_browser = epeek.genome();
genome_browser.start({ species : "human",
                       gene    : "BRCA2"
                     });
  </pre></code>
  <pre><code class="javascript">
var genome_browser = epeek.genome();
genome_browser.start({ species : "human",
                       chr     : 13,
                       from    : 139424940,
                       to      : 141784100
                     });
  </code></pre>

  <p>If the location is not given, the genome browser will start on the last location specified by the genome_browser.<a href="#species">species</a>, genome_browser.<a href="#chr">chr</a>, genome_browser.<a href="#from">from</a>, genome_browser.<a href="#to">to</a> and genome_browser.<a href="#gene">gene</a></p>
  <pre><code class="javascript">
var genome_browser = epeek.genome().species("human").chr("13").from(139424940).to(141784100);
genome_browser.start();
  </pre></code>
  <pre><code class="javascript">
var genome_browser = epeek.genome().species("human").gene("BRCA2");
genome_browser.start();
  </pre></code>

  <p>If a gene name is specified, the genome browser starts by trying to retrieve it from the Ensembl REST server. The gene name may be and <i>Ensembl ID</i> or a common name / external reference. Depending on which one is used, a different course of action is taken. See genome_browser.<a href="#gene">gene</a> for more details</p>

<!-- Controls -->
  <h2><a name="controls" href="#controls">#</a>Controls</h2>
  <p>Although panning and zoomming in the epeek.genome is normally driven by mouse gestures, the API also offers an alternative for API-driven controls. This may be useful in certain situations (see for example the <a href="../../../themes/comparative/comparative.html">comparative theme</a>). For an example of how to use these controls see the <a href="../../../themes/buttons/buttons.html">buttons theme.</a>

    <p><a name="allow_drag" href="#allow_drag">#</a> genome_browser.<b>allow_drag</b> (&lt;boolean&gt;)</p>
    <p>This method sets the response of the genome browser to mouse panning/zooming. By default, "draggability" on the genome browser is set to <i>true</i>. Setting the "draggability" to <i>false</i>, means that the genome browser won't react to direct mouse events on the genome browser area. See the <a href="../../../themes/buttons/buttons.html">buttons theme</a> for an example of usage. If no argument is provided, the current boolean value is returned. If the argument is present, the original epeek object is returned to allow method chaining</p>

    <p><a name="right" href="#right">#</a> genome_browser.<b>right</b> (&lt;float&gt;)</p>
    <p>This method pans the genome browser to the right. The <i>factor</i> argument specifies the amount of panning, i.e. 1.2 means 20% panning. The limit for panning right is the end of the chromosome / seq_region</p>

    <p><a name="left" href="#left">#</a> genome_browser.<b>left</b> (&lt;float&gt;)</p>
    <p>This method pans the genome browser to the left. The <i>factor</i> argument specifies the amount of panning, i.e. 1.2 means 20% panning. The limit for panning left is the beginning of the chromosome / seq_region </p>

    <p><a name="zoom" href="#zoom">#</a> genome_browser.<b>zoom</b> (&lt;float&gt;)</p>
    <p>This method zooms in or out the genome browser. The <i>factor</i> argument specifies the amount of panning, i.e. 1.2 means zoomming in 20% and 0.8 means zoomming out 20%. The limit for zoomming in is 200bp and the limit for zoomming out is currently imposed by the Ensembl REST API (see epeek.rest.<a href="../rest/index.html#limits">limits</a>)</p>

<!-- Aspect -->
    <h2><a name="aspect" href="#aspect">#</a>Aspect</h2>
    <p>Many aspects of how the epeek plug-in looks are configurable to allow themes to integrate into the style of the host page. All these methods can be set before the genome <a href="#start">starts</a>, while some of them can also be re-set lively, meaning that they can be changed at any time before or after the genome browser has started. See the methods below for details.</p>

    <p>Also, a basic way of inserting marks in the genome browser is supported via the genome_browser.<a href="#pin">pin</a> method (see below)</p>

    <p><a name="width" href="#width">#</a> genome_browser.<b>width</b> ([number])</p>
    <p>This method sets the width in pixels of the genome browser. If no argument is provided, the current width is returned. The argument should be only a number without any suffix (like "px"). The default width value is 600 pixels. A new width can be set before the genome browser <a href="#start">starts</a> or after. In the latter case, the genome browser is resized to the new width. See the <a href="../../themes/resize/resize.html">resize</a> theme for an example. If an argument is provided, the epeek object is returned to allow method chaining</p>

  <p><a name="height" href="#height">#</a> genome_browser.<b>height</b> ([number])</p>
  <p>This method sets the height in pixels of the genome browser. If no argument is provided, the current height is returned. The argument should be a number without any suffix (like "px"). The default height is 150 pixels. A new height can be set before the genome browser <a href="#start">starts</a> or after. In the latter case, the genome browser is resized to the new height. See the <a href="../../themes/resize/resize.html">resize</a> theme for an example. If an argument is provided, the epeek object is returned to allow method chaining</p>

  <p><a name="background_color" href="#background_color">#</a> genome_browser.<b>background_color</b> ([string])</p>
  <p>This method sets the background color for the genome browser. This color is used in different places and should contrast with the dark color used for the font. This means that for better results, the background color should be clear and the foreground color darker. The argument should be a valid hexadecimal number (including the "#" prefix, like "#FFFFFF"). If no argument is provided, returns the current background color. If an argument is provided, the epeek object is returned to allow method chaining</p>

  <p><a name="foreground_color" href="#foreground_color">#</a> genome_browser.<b>foreground_color</b> ([string])</p>
  <p>This method sets the foreground color for the genome browser. This color is used in different places and should contrast with the defined <a href="#background_color">background_color</a>. The argument should be a valid hexadecimal number (including the "#" prefix, like "#FFFFFF"). If no argument is provided, returns the current background color. If an argument is provided, this method returns the epeek object to allow method chaining.</p>

  <p><a name="pin" href="#pin">#</a> genome_browser.<b>pin</b> (&lt;array&gt;, [string])</p>
  <p><aside>See the <a href="../../../themes/pins/pins.html">pin theme</a> for an example of usage</aside></p>
  <p>This method allows to place markers (pins) in the genome browser. It expects an array containing genomic locations (i.e. an array of numbers). This method returns the path to the pin file. This is useful to create a legend explaining the pin. Every time the method is called, a different pin is placed and returned. Currently there are only 6 different pins, so once they are exhausted no more pins can be placed in the genome browser. An optional second parameter specifying the path to the pin file may be set. For example:</p>

  <pre><code class="javascript">
var gB = epeek.genome();
var pin1_url = gB.pin([x1, x2, x3]); // Will place "red" pins
var pin2_url = gB.pin([x4, x5]);     // Will place "blue" pins

// pin1_url and pin2_url can be used to create a legend explaining the marks

// To add more "red" pins:
gB.pin([x10, x11], pin1_url);
  </code></pre>

  <p><a name="ease" href="#ease">#</a> genome_browser.<b>ease</b> ([string])</p>
  <p>This method sets the <a href="http://www.robertpenner.com/easing/">easing function</a> for the transitions in the genome browser. Its argument is a string specifying a new easing function. See <a href="https://github.com/mbostock/d3/wiki/Transitions#wiki-d3_ease">d3.ease</a> for more details. The default easing function is "cubic-in-out". The method returns the original epeek object to allow method chaining.</p>

  <h2><a name="data_retrieval" href="#data_retrieval">#</a>Data retrieval</h2>
  <p>Different methods are provided to retrieve Ensembl data from the genome browser:</p>

  <p><a name="homologues" href="#homologues">#</a> genome_browser.<b>homologues</b> (&lt;string&gt;, [callback])</p>
  <p>This method retrieves homology information from the <a href="http://beta.rest.ensembl.org/">Ensembl REST API</a>. The first argument is the ensembl ID of the gene to look for. This argument is mandatory. The second argument is optional and corresponds with a callback function. This function is run after the homologues are retrieved from the REST server. An object containing all the homologues found is passed to this callaback as its argument. This object contains two fields: <i>orthologues</i> and <i>paralogues</i> and each of them contains an array of objects containing the following information:</p>
  <p>
    <ul>
      <li><b>id</b>         : The Ensembl Gene ID of the homolog</li>
      <li><b>protein_id</b> : The Ensembl Protein ID of the homolog</li>
      <li><b>species</b>    : The species name the homolog belongs to</li>
      <li><b>type</b>       : The type of homology</li>
      <li><b>subtype</b>    : The subtype of the homology relationship</li>
    </ul>
  </p>

  <p>For example:</p>
  <pre><code class="javascript">
var homologues_cbak = function(homologues_obj) {
   var orthologues = homologues_obj.orthologues; // An array of orthologues
   var paralogues  = homologues_obj.paralogues;  // An array of paralogues

   for (var i = 0; i < orthologues.length; i++) {
     var ortholog = orthologues[i];
     var ortholog_id = ortholog.id
     var ortholog_sp = ortholog.species;
     // do whatever you want...
   };
};
genome_browser.homologues(ensGene.id, homologues_cbak);
  </code></pre>

  <p><a name="genes" href="#genes">#</a> genome_browser.<b>genes</b> ()</p>
  <p>This method returns an array containing the set of genes that are currently displayed by the genome browser. Each of these genes is an object having the same fields described in genome_browser.<a href="#gene_info_callback">gene_info_callback</a></p>

  <h2><a name="location" href="#location">#</a>Location</h2>
  <p>These methods allow to set the original coordinates to <a href="#start">start</a> the genome browser. Once the genome browser has been initialized, they can also be used to retrieve the actual genomic location the genome browser is in. A default location is given to make sure that the genome browser doesn't break if you forget to provide one. The current default is between positions 139424940 and 141784100 of human chromosome 7</p>

  <p><a name="species" href="#species">#</a> genome_browser.<b>species</b> ([string])</p>
  <p>This method sets/gets the species used in the REST queries. If no argument is provided, returns the current species or the default one if no one has been set before. Common names are allowed (for example, "human", "chimp", "gorilla", "mouse", etc...). Binary scientific names are also allowed with or without underscores (for example, "mus musculus" and "mus_musculus" are both allowed). Case is ignored. If an argument is provided, returns the epeek object allowing method chaining</p>

  <p><a name="chr" href="#chr">#</a> genome_browser.<b>chr</b> ([string])</p>
  <p>This method sets/gets the chr (or more formally, the seq_region) used in the REST queries. If no argument is provided, returns the current seq_region or the default one if no one has been set before. Strictly speaking, the method expects a seq_region as argument, i.e. scaffolds or contig names are valid and considered as chromosomes. If an argument is provided, the method returns the epeek object allowing method chaining</p>

  <p><a name="from" href="#from">#</a> genome_browser.<b>from</b> ([number])</p>
  <p>This method gets/sets the start coordinate of the genome browser. If no argument is provided, returns the current start coordinate or the default one if none has been set explicitely. If present, the argument must be numeric. Commas and dots (for example 32,341,674 or 32.341.674) are not allowed. If an argument is provided, the method returns the epeek object allowing method chaining</p>

  <p><a name="to" href="#to">#</a> genome_browser.<b>to</b> ([number])</p>
  <p>This method gets/sets the end coordinate of the genome browser. If no argument is provided, returns the current end coordinate or the default one if none has been set explicitely. If present, the argument must be numeric. Commas and dots (for example 32,341,674 or 32.341.674) are not allowed. If an argument is provided, the method returns the epeek object allowing method chaining</p>

  <p><a name="gene" href="#gene">#</a> genome_browser.<b>gene</b> ([string])</p>
  <p>This method gets/sets the gene to be searched the next time the genome_browser.<a href="#start">start</a> method is called without a specific location. The gene can be an <i>Ensembl ID</i> or an external reference/name. This means that for BRCA1, you can use either "ENSGXXXXXXXXX" or "BRCA1" (provided that the genome_browser.<a href="#species">species</a> is set to <i>human</i>). Despite both cases look similar, the actions that are taken place under the hood when the genome_browser.<a href="#start">start</a> method is called are very different. In the first case (using an <i>ensembl ID</i>), the <a href="http://beta.rest.ensembl.org/documentation/info/symbol_lookup">symbol_lookup</a> endpoint of the REST API is queried. After the information associated with that gene is retrieved, the genome_browser.<a href="#ensgene_search_callback">ensgene_search_callback</a> is executed passing the gene information as its argument. In the latter case, the <a href="http://beta.rest.ensembl.org/documentation/info/xref_external">xrefs/symbol</a> endpoint of the Ensembl REST API is queried. This results in a (possibly empty) list containing the Ensembl IDs associated witht the searched term. At this point, the genome_browser.<a href="#xref_search_callback">xref_search_callback</a> is called with this list as its argument. The first element in the array is then used to query the <a href="http://beta.rest.ensembl.org/documentation/info/symbol_lookup">symbol_lookup</a> endpoint following the steps described above (i.e. when an ensembl ID is used in the search).</p>

  <h2><a name="callbacks" href="#callbacks">#</a>Callbacks</h2>
  <p>The epeek core plug-in API offers a variety of callbacks that are executed in response to different situations. For example, genome_browser.<i>gene_info_callback</i> is executed every time a gene is selected in the genome browser. Its default action is to do nothing, but you can set this callback to any thing you want like showing basic information of the gene in a tooltip. Another example is genome_browser.<i>genes_callback</i> that is fired every time the REST server is queried as a result of a drag/pan event. Details of each of them are shown below</p>
  <p>Since the default action for these callbacks is to do nothing, in order to use them for real purposes they have to be redefined, normally in a theme. For example:</p>
  <pre><code class="javascript">
genome_browser.gene_info_callback = function (obj) {
   // The code inserted here will be executed every time a
   // gene is clicked
};
  </code></pre>

  <p><a name="gene_info_callback" href="#gene_info_callback">#</a> genome_browser.<b>gene_info_callback</b> (&lt;object&gt;)</p>
  <p>This callback is executed as a response to a gene being selected (i.e. "clicked") in the genome browser. The callback gets an object with the gene information as its argument. This object has the following fields:</p>

  <p>
    <ul>
      <li><b>external_name</b>   : The external name of the gene</li>
      <li><b>ID</b>              : The Ensembl ID of the gene</li>
      <li><b>description</b>     : A short description of the gene</li>
      <li><b>logic_name</b>      : The source of the gene</li>
      <li><b>seq_region_name</b> : The name of the chromosome or seq_region the gene is located</li>
      <li><b>start</b>           : The start coordinate of the gene in the chromosome / seq_region (start &lt; end)</li>
      <li><b>end</b>             : The end coordinate of the gene in the chromosome / seq_region (start &lt; end)</li>
      <li><b>strand</b>          : The strand of the gene</li>
    </ul>
  </p>

  <pre><code class="javascript">
genome_browser.gene_info_callback = function (obj) {
      // obj.external_name...
};
  </code></pre>

  <p>It is possible to assign the <i>gene_info_callback</i> to <a href="../tooltip/index.html">epeek.tooltip</a>. Doing so, a tooltip showing basic information will appear every time a gene is selected. See genome.<a href="#tooltip">tooltip</a> and the <a href="">tooltip theme</a> for details.</p>

  <pre><code class="javascript">
gB.gene_info_callback = gB.tooltip();
  </code></pre>

  <p><a name="xref_search_callback" href="#xref_search_callback">#</a> genome_browser.<b>xref_search_callback</b> (&lt;array&gt;)</p>
  <p>This callback is called every time an external name is searched in the REST server. Although it may seem that this method and <a href="#ensgene_search_callback" >ensgene_search_callback</a> may overlap, they are executed in very different situations. This method is called every time the <a href="http://beta.rest.ensembl.org/documentation/info/xref_external">xrefs/symbol</a> endpoint is queried, i.e. every time we need to find the ensembl IDs associated with a given common name or xref. This may result in more than one entry being returned by the REST server. The latter is called every time a gene is queried by ensembl ID, i.e. we know the gene we want because we have its ID. See the genome_browser.<a href="#gene">gene</a> method for details.</p>
  <p>The callback is called passing an array of objects having the following fields:</p>

  <p>
    <ul>
      <li><b>id</b>   : The Ensembl gene id associated with the xref</li>
      <li><b>type</b> : The type of object the ID refers to (this should always be "gene")</li>
    </ul>
  </p>

  <p>For example:</p>

  <pre><code class="javascript">
genome_browser.gene_search_callback = function(arr) {
   for (var i = 0; i < arr.length; i++) {
      // arr[i].id has the ensembl ID of each gene associated with the xrefs.
   }
};
  </code></pre>

  <p><a name="genes_callback" href="#genes_callback">#</a> genome_browser.<b>genes_callback</b> (&lt;array&gt;)</p>
  <p>This callback is called every time a list of genes are returned from the REST server as a result of a pan/drag/zoom event. The callback is called passing as its argument an array of objects having the same fields described in genome_browser.<a href="#gene_info_callback">gene_info_callback</a>.</p>

  <p><a name="tooltip" href="#tooltip">#</a> genome_browser.<b>tooltip</b> ()</p>
  <p>This method returns a closure that can be used to display gene information as a tooltip using the <a href="../tooltip/index.html">epeek.tooltip</a> interface. A common way of using this method is by applying it to genome_browser.<a href="#gene_info_callback">gene_info_callback</a>. By doing so, every time a gene is <i>clicked</i> in the genome browser, a tooltip showing basic gene information will appear:</p>

  <pre><code class="javascript">
var gB = epeek.genome();
gB.gene_info_callback = gB.tooltip();
  </code></pre>

  <p>See the <a href="../../../themes/tooltips/tooltips.html">tooltip theme</a> for a complete example</p>

  <footer>
    <aside>Released under <a href="http://www.opensource.org/licenses/apache2.0.php">Apache 2.0</a>.</aside>
    Copyright 2014 <a href="http://www.ebi.ac.uk/~mp" rel="author">Miguel Pignatelli</a>
  </footer>

</div>

<script src="../../../highlight.v0.min.js"></script>

